### ISSUE TITLE
solana: OldGuardianSet error on second guardian set upgrade

### ISSUE THREAD

**Issue by leoluk** [certusone/wormhole#110]
Reproduction instructions on #104:

- Launch clean devnet with `--num=1`.
- Migrate guardian set from 0 -> 1 via `scripts/test-injection.sh 0 0`.
- Migrate guardian set from 1 -> 2 via `scripts/test-injection.sh 0 1`.

Log: https://cloud.tilt.dev/snapshot/AcitrvsLUFCCL548DBI=

Both guardian set updates are accepted by the Ethereum contract:

```
[guardiand] 2020-11-19T01:29:52.490Z	INFO	wormhole-guardian-0.root	lockup submitted to Ethereum	{"tx": {"nonce":"0x3","gasPrice":"0x4a817c800","gas":"0x2348f","to":"0x5b1869d9a4c187f2eaa108f3062412ecf0526b24","value":"0x0","input":"0x3bc0aee60000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000006601000000000100a4ee1a7534cc74617a8ea4bde9711a1a6593c37a660ea356b11e3b77ed79dd324437610dfa8c19074c2ccb04c19d3150694395412d5e33e597ef9a24e95d273e005fb5b7a1010000000101befa429d57cd18b7f8a4d91a2da9ab4af05d0fbe0000000000000000000000000000000000000000000000000000","v":"0x1c","r":"0x7227773e4456264f4de66d81e73ad4c41f3c8c88e2c094fd9d32a5a02565a0bf","s":"0x1f752cc6a943315cc2ae352546b50cb968f711f9b4c4cbf2517c518cb3448db5","hash":"0xf1e707d6b0c30e4c44df6dd126bb7013299a42fba1ce227095ab10295f756176"}, "digest": "698b6be79edfbae1e2f4097bbc72019d78672d2c3e3706809ab3757870ab4c19"}
[guardiand] 2020-11-19T01:29:52.500Z	INFO	wormhole-guardian-0.root.ethwatch	guardian set has changed, fetching new value	{"new_index": 1}
```

```
[guardiand] 2020-11-19T01:30:11.714Z	INFO	wormhole-guardian-0.root	lockup submitted to Ethereum	{"tx": {"nonce":"0x4","gasPrice":"0x4a817c800","gas":"0x2349b","to":"0x5b1869d9a4c187f2eaa108f3062412ecf0526b24","value":"0x0","input":"0x3bc0aee600000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000066010000000101001a128500516c61f8e883bc7b042c9393a5d3747df56d1277d422e5d3584cf32b127b0e3183bf7fe4e1630763a045c6c9160a0ab6da8b86dfca1fa350bb889ac0015fb5b7a1010000000201befa429d57cd18b7f8a4d91a2da9ab4af05d0fbe0000000000000000000000000000000000000000000000000000","v":"0x1b","r":"0xb393654a4929bd8e72e2c74b8bce17013cbce39d35b8d98e08da3a20804abe85","s":"0x46ca89f6a99ae8d4322f142ccf160303c43785c20a7e17e341d972b0e549313d","hash":"0x3f03b34b1a1dfd72f6ab29e559922d71dfc7cd1cd4da9c2dd576cf880d8c010d"}, "digest": "9a058cac33efac9cb304f2c2da49cab52b0984f1ec8717dfe6a599aa80689bc8"}
[guardiand] 2020-11-19T01:30:11.718Z	INFO	wormhole-guardian-0.root.ethwatch	guardian set has changed, fetching new value	{"new_index": 2}
```

The first guardian set update is accepted by Solana:

```
[guardiand] 2020-11-19T01:29:51.739Z	INFO	wormhole-guardian-0.root	submitting signed VAA to Solana	{"digest": "698b6be79edfbae1e2f4097bbc72019d78672d2c3e3706809ab3757870ab4c19", "vaa": {"Version":1,"GuardianSetIndex":0,"Signatures":[{"Index":0,"Signature":[164,238,26,117,52,204,116,97,122,142,164,189,233,113,26,26,101,147,195,122,102,14,163,86,177,30,59,119,237,121,221,50,68,55,97,13,250,140,25,7,76,44,203,4,193,157,49,80,105,67,149,65,45,94,51,229,151,239,154,36,233,93,39,62,0]}],"Timestamp":"2020-11-19T00:09:05Z","Payload":{"Keys":["0xbefa429d57cd18b7f8a4d91a2da9ab4af05d0fbe"],"NewIndex":1}}, "bytes": "01000000000100a4ee1a7534cc74617a8ea4bde9711a1a6593c37a660ea356b11e3b77ed79dd324437610dfa8c19074c2ccb04c19d3150694395412d5e33e597ef9a24e95d273e005fb5b7a1010000000101befa429d57cd18b7f8a4d91a2da9ab4af05d0fbe"}
[guardiand] 2020-11-19T01:29:53.758Z	INFO	wormhole-guardian-0.root.solwatch	submitted VAA	{"tx_sig": "4szGgVr5LT5yx3u6e45pBxsLRaugJ5tBhoGNKWQFJyaJfvNgnQRFAaN7i2SGsq5fyMza4sXsRg7cUGddW3WgzKTe", "digest": "698b6be79edfbae1e2f4097bbc72019d78672d2c3e3706809ab3757870ab4c19"}
```

...but the second one fails:

```
[guardiand] 2020-11-19T01:30:11.259Z	INFO	wormhole-guardian-0.root	submitting signed VAA to Solana	{"digest": "9a058cac33efac9cb304f2c2da49cab52b0984f1ec8717dfe6a599aa80689bc8", "vaa": {"Version":1,"GuardianSetIndex":1,"Signatures":[{"Index":0,"Signature":[26,18,133,0,81,108,97,248,232,131,188,123,4,44,147,147,165,211,116,125,245,109,18,119,212,34,229,211,88,76,243,43,18,123,14,49,131,191,127,228,225,99,7,99,160,69,198,201,22,10,10,182,218,139,134,223,202,31,163,80,187,136,154,192,1]}],"Timestamp":"2020-11-19T00:09:05Z","Payload":{"Keys":["0xbefa429d57cd18b7f8a4d91a2da9ab4af05d0fbe"],"NewIndex":2}}, "bytes": "010000000101001a128500516c61f8e883bc7b042c9393a5d3747df56d1277d422e5d3584cf32b127b0e3183bf7fe4e1630763a045c6c9160a0ab6da8b86dfca1fa350bb889ac0015fb5b7a1010000000201befa429d57cd18b7f8a4d91a2da9ab4af05d0fbe"}
[guardiand] 2020-11-19T01:30:12.273Z	ERROR	wormhole-guardian-0.root.solwatch	error submitting VAA	{"error": "rpc error: code = Internal desc = tx sending failed: RPC response error -32002: Transaction simulation failed: Error processing Instruction 0: custom program error: 0x14 [7 log messages]", "digest": "9a058cac33efac9cb304f2c2da49cab52b0984f1ec8717dfe6a599aa80689bc8"}
```

```
[devnet] [2020-11-19T01:30:12.270004332Z INFO  solana_runtime::message_processor] Call BPF program Bridge1p5gheXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o
[devnet] [2020-11-19T01:30:12.270142363Z INFO  solana_runtime::message_processor] Program log: Instruction: PostVAA
[devnet] [2020-11-19T01:30:12.270699208Z INFO  solana_runtime::message_processor] Program log: deriving key
[devnet] [2020-11-19T01:30:12.270815379Z INFO  solana_runtime::message_processor] Program log: deploying contract
[devnet] [2020-11-19T01:30:12.271011380Z TRACE solana_runtime::system_instruction_processor] process_instruction: CreateAccount { lamports: 1169280, space: 40, owner: Bridge1p5gheXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o }
[devnet] [2020-11-19T01:30:12.271031640Z TRACE solana_runtime::system_instruction_processor] keyed_accounts: [KeyedAccount { is_signer: true, is_writable: true, key: 6sbzC1eH4FTujJXWj51eQe25cYvr4xfXbJ1vAj7j2k5J, account: RefCell { value: Account { lamports: 99969535880 data.len: 0 owner: 11111111111111111111111111111111 executable: false rent_epoch: 0 } } }, KeyedAccount { is_signer: true, is_writable: true, key: GJdAaZXkkvmBpqP5WdwQwaWatoAUa69yirrJbcLTvRD3, account: RefCell { value: Account { lamports: 0 data.len: 0 owner: 11111111111111111111111111111111 executable: false rent_epoch: 0 } } }]
[devnet] [2020-11-19T01:30:12.271790637Z INFO  solana_runtime::message_processor] Program log: Error: OldGuardianSet
[devnet] [2020-11-19T01:30:12.271848727Z DEBUG solana_rbpf::vm] BPF instructions executed: 59849
[devnet] [2020-11-19T01:30:12.271852457Z DEBUG solana_rbpf::vm] Max call depth reached: 15
[devnet] [2020-11-19T01:30:12.271857487Z INFO  solana_runtime::message_processor] BPF program consumed 65749 of 200000 units
[devnet] [2020-11-19T01:30:12.271866347Z INFO  solana_runtime::message_processor] BPF program Bridge1p5gheXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o failed: custom program error: 0x14
```

**Comment by leoluk** [comment]
```diff
--- solana/bridge/src/processor.rs	(revision 25fad01794e5fbaaae2bf0009a566cf98f48bbc3)
+++ solana/bridge/src/processor.rs	(date 1605750772462)
@@ -700,6 +700,8 @@
     ) -> ProgramResult {
         let new_guardian_info = next_account_info(account_info_iter)?;
 
+        info!(0x1000, bridge.guardian_set_index, old_guardian_set.index, 0, 0);
+
         // The new guardian set must be signed by the current one
         if bridge.guardian_set_index != old_guardian_set.index {
             return Err(Error::OldGuardianSet.into());
@@ -757,6 +759,8 @@
         // Update the bridge guardian set id
         bridge.guardian_set_index = b.new_index;
 
+        info!(0x1001, bridge.guardian_set_index, b.new_index, 0, 0);
+
         Ok(())
     }
 ```

```
[devnet] [2020-11-19T01:54:57.389674563Z INFO  solana_runtime::message_processor] Program log: 0x1000, 0x0, 0x0, 0x0, 0x0
[devnet] [2020-11-19T01:54:57.389946125Z INFO  solana_runtime::message_processor] Program log: 0x1001, 0x1, 0x1, 0x0, 0x0
[devnet] [2020-11-19T01:55:03.212466089Z INFO  solana_runtime::message_processor] Program log: 0x1000, 0x0, 0x1, 0x0, 0x0
[devnet] [2020-11-19T01:55:03.212481239Z INFO  solana_runtime::message_processor] Program log: Error: OldGuardianSet
```

**Comment by leoluk** [comment]
Fixed via #82
